keystoremanager/keystoremanager.go

```diff
@@ -168,172 +183,239 @@ func (km *KeyStoreManager) GetRootCryptoService(rootKeyID, passphrase string) (*
+	logrus.Debugf("entering root certificate rotation for: %s", gun)
+
+	// Do root certificate rotation: we trust only the certs present in the new root
+	// First we add all the new certificates (even if they already exist)
+	for _, cert := range allValidCerts {
+		err := km.trustedCertificateStore.AddCert(cert)
+		if err != nil {
+			// If the error is already exists we don't fail the rotation
+			if _, ok := err.(*trustmanager.ErrCertExists); ok {
+				continue
+			}
+			logrus.Debugf("error adding new trusted certificate for: %s, %v", gun, err)
+		}
+		// Getting the CertID for debug only, don't care about the error
+		certID, _ := trustmanager.FingerprintCert(cert)
+		logrus.Debugf("adding trust certificate with certID %s for %s", certID, gun)
+	}
+
+	// Now we delete old certificates that aren't present in the new root
+	for certID, cert := range certsToRemove(certsForCN, allValidCerts) {
+		logrus.Debugf("removing certificate with certID: %s", certID)
+		err = km.trustedCertificateStore.RemoveCert(cert)
+		if err != nil {
+			logrus.Debugf("failed to remove trusted certificate with keyID: %s, %v", certID, err)
+			return &ErrRootRotationFail{Reason: "failed to rotate root keys"}
+		}
+	}
+
+	logrus.Debugf("Root validation succeeded for %s", gun)
+	return nil
+}
+
+// validRootLeafCerts returns a list of non-exipired, non-sha1 certificates whoose
+// Common-Names match the provided GUN
+func validRootLeafCerts(root *data.SignedRoot, gun string) ([]*x509.Certificate, error) {
+	// Get a list of all of the leaf certificates present in root
+	allLeafCerts, _ := parseAllCerts(root)
+	var validLeafCerts []*x509.Certificate
+
+	// Go through every leaf certificate and check that the CN matches the gun
+	for _, cert := range allLeafCerts {
+		// Validate that this leaf certificate has a CN that matches the exact gun
+		if cert.Subject.CommonName != gun {
+			logrus.Debugf("error leaf certificate CN: %s doesn't match the given GUN: %s", cert.Subject.CommonName)
+			continue
+		}
+		// Make sure the certificate is not expired
+		if time.Now().After(cert.NotAfter) {
+			logrus.Debugf("error leaf certificate is expired")
+			continue
+		}
+
+		// We don't allow root certificates that use SHA1
+		if cert.SignatureAlgorithm == x509.SHA1WithRSA ||
+			cert.SignatureAlgorithm == x509.DSAWithSHA1 ||
+			cert.SignatureAlgorithm == x509.ECDSAWithSHA1 {
+
+			logrus.Debugf("error certificate uses old ciphers")
+			continue
+		}
+
+		validLeafCerts = append(validLeafCerts, cert)
+	}
+
+	if len(validLeafCerts) < 1 {
+		return nil, errors.New("no valid leaf certificates found in any of the root keys")
+	}
+
+	return validLeafCerts, nil
+}
 
-	// allCerts will keep a list of all leafCerts that were found, and is used
-	// to aid on root certificate rotation
-	allCerts := make(map[string]*x509.Certificate)
+func parseAllCerts(signedRoot *data.SignedRoot) (map[string]*x509.Certificate, map[string][]*x509.Certificate) {
```

Comment:
GetIntermediateCerts is used on line 380:

```go
intermediateCerts := trustmanager.GetIntermediateCerts(decodedCerts)
```

And it is correct that right now we don't use the second return argument of parsAllCerts, because we are not supporting CAs.